% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/feature_level_diagnostics.R
\name{feature_level_diagnostics}
\alias{feature_level_diagnostics}
\alias{plot_single_feature}
\title{Plot peptide measurements}
\usage{
plot_single_feature(feature_name, df_long, sample_annotation,
  sample_id_col = "FullRunName", measure_col = "Intensity",
  feature_id_col = "peptide_group_label", geom = c("point", "line"),
  color_by_col = NULL, color_by_value = NULL, batch_col = "MS_batch",
  color_by_batch = FALSE, color_scheme = "brewer",
  order_col = "order", vline_color = "red", facet_col = NULL,
  plot_title = NULL, theme = "classic", ylimits = NULL)
}
\arguments{
\item{feature_name}{name of the selected feature (e.g. peptide) for diagnostic profiling}

\item{df_long}{data frame where each row is a single feature in a single
sample. It minimally has a \code{sample_id_col}, a \code{feature_id_col} and a
\code{measure_col}, but usually also an \code{m_score} (in OpenSWATH output result
file)}

\item{sample_annotation}{data frame with: \enumerate{ \item \code{sample_id_col}
(this can be repeated as row names) \item biological covariates \item
technical covariates (batches etc) }}

\item{sample_id_col}{name of the column in \code{sample_annotation} table, where the
filenames (colnames of the data matrix are found).}

\item{measure_col}{if \code{df_long} is among the parameters, it is the
column with expression/abundance/intensity; otherwise, it is used
internally for consistency.}

\item{feature_id_col}{name of the column with feature/gene/peptide/protein
ID used in the long format representation \code{df_long}. In the wide
formatted representation \code{data_matrix} this corresponds to the row
names.}

\item{geom}{whether to show the feature as points and/or connect by lines (accepted 
values are: 1. \code{point}, \code{line} and \code{c('point', 'line')})}

\item{color_by_col}{column to color point by certain value denoted 
by \code{color_by_value}. Design with inferred/requant values in openSWATH output data, 
which means argument value has to be set to `m_score`.}

\item{color_by_value}{value in \code{color_by_col} to color. For OpenSWATH data,
this argument value has to be set to `2` (this is an `m_score` value for requants).}

\item{batch_col}{column in \code{sample_annotation} that should be used for
batch comparison}

\item{color_by_batch}{(logical) whether to color points and connecting lines 
by batch factor as defined by \code{batch_col}.}

\item{order_col}{column in \code{sample_annotation} that determines sample order. It is
used for in initial assessment plots (\link{plot_sample_mean_or_boxplot}) and 
feature-level diagnostics (\link{feature_level_diagnostics}). Can be `NULL` 
if sample order is irrelevant (e.g. in genomic experiments). For more details,
order definition/inference, see \link{define_sample_order} and \link{date_to_sample_order}}

\item{vline_color}{color of vertical lines, typically separating 
different MS batches in ordered runs; 
should be `NULL` for experiments without intrinsic order.}

\item{facet_col}{column  in `sample_annotation` with a batch factor to separate 
plots into facets; usually 2nd to `batch_col`. Most meaningful for multi-instrument 
MS experiments (where each instrument has its own order-associated effects) 
or simultaneous examination of two batch factors (e.g. preparation day and measurement day)}

\item{plot_title}{the string indicating the source of the peptides}

\item{theme}{plot theme (default is 'classical'; other options not
implemented)}

\item{ylimits}{range of y-axis to plot feature-level trends}
}
\value{
ggplot2 type plot of \code{measure_col} vs \code{order_col},
  faceted by \code{pep_name} and (optionally) by \code{batch_col}
}
\description{
Creates a peptide facetted ggplot2 plot of the value in \code{measure_col}
vs \code{order_col} (if `NULL`, x-axis is simply a sample name order). 
Additionally, the resulting plot can also be facetted by a batch factor, 
e.g. an MS instrument.
}
\examples{
plot_single_feature(pep_name = "46213_NVGVSFYADKPEVTQEQK_2", 
df_long = example_proteome, example_sample_annotation, 
color_by_col = NULL)

#to examine features that have missing values specific per batch, this can be used:
plot_single_feature(pep_name = "46213_NVGVSFYADKPEVTQEQK_2", 
df_long = example_proteome, example_sample_annotation, 
color_by_col = 'm_score', color_by_value = 2)

}
\seealso{
Other feature-level diagnostic functions: \code{\link{plot_iRT}},
  \code{\link{plot_peptides_of_one_protein}},
  \code{\link{plot_spike_in}},
  \code{\link{plot_with_fitting_curve}}
}
\concept{feature-level diagnostic functions}
